coral-incremental/src/main/java/com/linkedin/coral/incremental/RelNodeIncrementalTransformer.java

```diff
@@ -49,61 +66,130 @@ public RelNode visit(TableScan scan) {
+        incrementalTransformerResults.addIntermediateQueryRelNode(getTableNameFromDescription(project) + "_delta",
+            transformedProject);
+        return transformedProject;
       }
 
       @Override
       public RelNode visit(LogicalUnion union) {
         List<RelNode> children = union.getInputs();
-        List<RelNode> transformedChildren =
+        List<IncrementalTransformerResults> incrementalTransformerResultsChildren =
             children.stream().map(child -> convertRelIncremental(child)).collect(Collectors.toList());
+        List<RelNode> transformedChildren = new ArrayList<>();
+        for (IncrementalTransformerResults incrementalTransformerResultsChild : incrementalTransformerResultsChildren) {
+          transformedChildren.add(incrementalTransformerResultsChild.getIncrementalRelNode());
+          incrementalTransformerResults
+              .addMultipleIntermediateQueryRelNodes(incrementalTransformerResultsChild.getIntermediateQueryRelNodes());
+        }
         return LogicalUnion.create(transformedChildren, union.all);
       }
 
       @Override
       public RelNode visit(LogicalAggregate aggregate) {
-        RelNode transformedChild = convertRelIncremental(aggregate.getInput());
+        IncrementalTransformerResults incrementalTransformerResultsChild = convertRelIncremental(aggregate.getInput());
+        RelNode transformedChild = incrementalTransformerResultsChild.getIncrementalRelNode();
+        incrementalTransformerResults
+            .addMultipleIntermediateQueryRelNodes(incrementalTransformerResultsChild.getIntermediateQueryRelNodes());
         return LogicalAggregate.create(transformedChild, aggregate.getGroupSet(), aggregate.getGroupSets(),
             aggregate.getAggCallList());
       }
     };
-    return originalNode.accept(converter);
+    incrementalTransformerResults.setIncrementalRelNode(originalNode.accept(converter));
+    return incrementalTransformerResults;
   }
 
-  private static LogicalProject createProjectOverJoin(LogicalJoin join, RelNode left, RelNode right,
+  private static String getTableNameFromDescription(RelNode relNode) {
+    String identifier = relNode.getDescription().split("#")[1];
+    return "Table#" + identifier;
+  }
+
+  private static LogicalProject createReplacementProjectNodeForGivenRelNode(String relOptTableName, RelNode relNode,
       RexBuilder rexBuilder) {
-    LogicalJoin incrementalJoin =
-        LogicalJoin.create(left, right, join.getCondition(), join.getVariablesSet(), join.getJoinType());
+    RelOptTable table =
+        RelOptTableImpl.create(relOptSchema, relNode.getRowType(), Collections.singletonList(relOptTableName), null);
+    TableScan scan = LogicalTableScan.create(relNode.getCluster(), table);
+    return createProjectOverNode(scan, rexBuilder);
+  }
+
+  private static LogicalProject createProjectOverNode(RelNode relNode, RexBuilder rexBuilder) {
```

Comment:
How about making this `createJoin`, and having it construct both he join and projection node? That is semantically closer to what we are trying to achieve.